Added custom keymap for preonic

elisiano opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description

<!--- Describe your changes in detail here. -->

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

fauxpark commented on keyboards/preonic/keymaps/elisiano/keymap.c:
```diff
@@ -0,0 +1,334 @@
+          break;
+        case EMOJI:
+          if (record->event.pressed) {
+            SEND_STRING(SS_DOWN(X_LCTRL)SS_DOWN(X_LGUI)SS_TAP(X_SPACE)SS_UP(X_LGUI)SS_UP(X_LCTRL));
```
This can be simplified a bit, and probably saves a handful of bytes:
```suggestion
            tap_code16(C(G(KC_SPC)));
```

fauxpark commented on keyboards/preonic/keymaps/elisiano/keymap.c:
```diff
@@ -0,0 +1,334 @@
+    }
+  } else {
+    if (clockwise) {
+      register_code(KC_PGDN);
```
These can be simplified too.
```suggestion
      tap_code(KC_PGDN);
```

fauxpark commented on keyboards/preonic/keymaps/elisiano/keymap.c:
```diff
@@ -0,0 +1,334 @@
+  } else {
+    if (clockwise) {
+      register_code(KC_PGDN);
+      unregister_code(KC_PGDN);
```
```suggestion
```

fauxpark commented on keyboards/preonic/keymaps/elisiano/keymap.c:
```diff
@@ -0,0 +1,334 @@
+      register_code(KC_PGDN);
+      unregister_code(KC_PGDN);
+    } else {
+      register_code(KC_PGUP);
```
```suggestion
      tap_code(KC_PGUP);
```

fauxpark commented on keyboards/preonic/keymaps/elisiano/keymap.c:
```diff
@@ -0,0 +1,334 @@
+      unregister_code(KC_PGDN);
+    } else {
+      register_code(KC_PGUP);
+      unregister_code(KC_PGUP);
```
```suggestion
```

fauxpark commented on keyboards/preonic/keymaps/elisiano/keymap.c:
```diff
@@ -0,0 +1,334 @@
+  LCTL_T(KC_ESC),   KC_A,    KC_S,    KC_D,    KC_F,    KC_G,    KC_H,    KC_J,    KC_K,    KC_L,    KC_SCLN, KC_QUOT, \
+  KC_LSFT,          KC_Z,    KC_X,    KC_C,    KC_V,    KC_B,    KC_N,    KC_M,    KC_COMM, KC_DOT,  KC_SLSH, KC_ENT,  \
+  EMOJI,            KC_LCTL, KC_LALT, KC_LGUI, LOWER,   KC_SPC,  KC_SPC,  RAISE,   KC_LEFT, KC_DOWN, KC_UP,   KC_RGHT  \
+),
```
The backslashes at the end of these lines are not necessary. You can keep them if you want as it isn't hurting anything, but IMO it looks neater without them.

fauxpark commented on keyboards/preonic/keymaps/elisiano/keymap.c:
```diff
@@ -0,0 +1,334 @@
+  DVORAK,
+  LOWER,
+  RAISE,
+  BACKLIT,
```
```suggestion
```
Since you're not using this keycode.

fauxpark commented on keyboards/preonic/keymaps/elisiano/keymap.c:
```diff
@@ -0,0 +1,334 @@
+            #endif
+          }
+          return false;
+          break;
```
This whole block can be deleted.

elisiano commented:
Apologies for the delay, this notification went buried in my mailbox.
I think I've addressed the comments, please let me know if there's anything else.
(I did flash the firmware and I'm typing with it right now 👍).

The reason why there were "extra" thing in this PR is because I was trying to keep my file in sync with master/default and apply only my changes. I guess it's not necessary 😀

fauxpark commented on keyboards/preonic/keymaps/elisiano/keymap.c:
```diff
@@ -0,0 +1,315 @@
+          break;
+        case EMOJI:
+          if (record->event.pressed) {
+            //SEND_STRING(SS_DOWN(X_LCTRL)SS_DOWN(X_LGUI)SS_TAP(X_SPACE)SS_UP(X_LGUI)SS_UP(X_LCTRL));
```
```suggestion
```

drashna commented on keyboards/preonic/keymaps/elisiano/keymap.c:
```diff
@@ -0,0 +1,314 @@
+            set_single_persistent_default_layer(_MOUSE);
+          }
+          return false;
+          break;
```
I'm not sure that you want this.  Having a `TO` keycode to get there and get back mau be closer to what you want.  

As is, this will stay, if you unplug or reset the keyboard.  If that's fine, then leave as is.  Otherwise, you may want to change this.

elisiano replied:
As I’m not familiar with qmk, I was trying to follow how other things are done. 
So if you know a better way of doing the same please let me know (in details though, because I’m missing context). 

That said, this works. I can switch to the Mouse layer and back to another without issues.

elisiano replied:
@drashna do you have any pointers on how to improve that?
Although I'm curious to understand why you said this will stay as in fact I can switch back to any other adjustment layer (with `LOWER+RISE+<KEY>`).

drashna replied:
Well, the best answers are in the documentation. 

https://docs.qmk.fm/#/ref_functions
https://docs.qmk.fm/#/feature_advanced_keycodes?id=switching-and-toggling-layers

ridingqwerty replied:
@elisiano: I think what @drashna was getting at here (correct me if I'm wrong) is that by using `set_single_persistent_default_layer(_MOUSE);`, you are changing the default layer and storing that to EEPROM, meaning if the board loses power, that would be your layer when plugging back in, which could be confusing.

You could replace the `MOUSE` keycode with `TO(_MOUSE)` and the `QWERTY` keycode on the `_MOUSE` layer with `TO(_QWERTY)` to do this is a way that doesn't involve the EEPROM.

That said, if this works for you as is, I see no reason to block merging this as it looks good besides what drashna pointed out.

drashna replied:
^ This is exactly what I meant.  

Also, `TG(_MOUSE)` would work too.

elisiano replied:
I see, thanks @ridingqwerty for the detailed explanation.
@drashna I will update the PR as soon as I get a chance!

elisiano replied:
I think I've addressed this, please let me know if there is anything else I need to look into.

ridingqwerty commented:
Thanks @elisiano , I think we're good here!
